docs/enhancements/transactional-custom-manifests.md

```diff
@@ -0,0 +1,260 @@
+
+To mark a file as user generated, a metadata file is created under the subtree `manifest-attributes` to match the file and path in the subtree `manifests`
+
+So if we have a file:
+`manifests/openshift/user-defined-manifest.yaml`
+
+The metadata file to mark this as user-supplied would be:
+`manifest-attributes/openshift/user-defined-manifest.yaml/user-supplied`
+
+As we are using the filesystem as a means of storing metadata about a file, this needs to be performed in two operations, one to write the file and one to write the metadata.
+
+This has a weakness; by performing two writes, we risk that the second write may not take place in the case of an S3 failure. 
+
+# Summary
+
+We would like to improve the custom manifest feature to make it more consistent, more reliable and more user friendly. By addressing the following points
+
+* Find a more reliable and "transactional" way in which the upload of manifest files may be handled.
+
+# Motivation
+
+We have had a recent bug report about an edge case
+
+  The following bug was contributed to by this issue: https://issues.redhat.com/browse/MGMT-17802
+
+In this edge case, if there is a failure in S3 with unfortunate timing then it is possible to store a manifest with inconsistent metadata.
+This is not a situation that is acceptable moving forward and we need to do something.
+
+There are some problems with the existing implementation of manifest storage when it comes to metadata
+
+* A manifest upload cannot be "transactional" under the current design, raising the risk that one part of the upload may fail and leave the rest inconsistent.
+
+### Goals
+
+- To ensure that manifests may be written atomically so that their reliability may be guaranteed.
+- To simplify the storage of manifest metadata.
+
+### Non-Goals
+
+- It is a non-goal to change the behavior of the UI for manifests.
+
+## Proposal: Use the Amazon S3 metadata API 
+
+We should change the way that we store metadata about a manifest, instead of using an awkward "directory based" approach to metadata, we should instead use the built in facilities of S3 which has native Metadata support for objects.
+This should be achievable in a single write and should mitigate the transactionality issues.
+
+### Object store API changes required:
+
+	#### Uploads
+
+	S3Wrapper API:
+	Calls to Upload, UploadFile and UploadStream will need to be updated to support metadata for an object
+	https://github.com/openshift/assisted-service/blob/master/pkg/s3wrapper/client.go#L42-L43
+	
+	##### Filesystem implementation:
+	
+	Implementation changes to the filesystem S3Wrapper client will be required in order to implement metadata support.
+	We should attempt to keep this atomic by using XATTRs if supported.
+	
+	Implementation of Upload in S3Wrapper API (filesystem):
+	Need to be updated to support metadata for an object (possibly based on XATTRs
+	https://github.com/openshift/assisted-service/blob/master/pkg/s3wrapper/filesystem.go#L55-L70
```

Comment:
FWIW these links will be inaccurate very soon since they reference `master`.

It would be better to use a commit for these to capture the state of the code when the enhancement was written (generally this also applies to everywhere else you put a GH link like emails, IMs, and comments).